k9mail/src/main/java/com/fsck/k9/activity/compose/AttachmentPresenter.java

```diff
@@ -0,0 +1,371 @@
+                initAttachmentInfoLoader(attachment);
+            } else if (attachment.state == LoadingState.METADATA) {
+                initAttachmentContentLoader(attachment);
+            }
+        }
+    }
+
+    public boolean checkOkForSendingOrDraftSaving() {
+        if (actionToPerformAfterWaiting != WaitingAction.NONE) {
+            return true;
+        }
+
+        if (hasLoadingAttachments()) {
+            actionToPerformAfterWaiting = WaitingAction.SEND;
+            attachmentMvpView.showWaitingForAttachmentDialog(actionToPerformAfterWaiting);
+            return true;
+        }
+
+        return false;
+    }
+
+    private boolean hasLoadingAttachments() {
+        for (Attachment attachment : attachments.values()) {
+            Loader loader = loaderManager.getLoader(attachment.loaderId);
+            if (loader != null && loader.isStarted()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public ArrayList<Attachment> createAttachmentList() {
+        ArrayList<Attachment> result = new ArrayList<>();
+        for (Attachment attachment : attachments.values()) {
+            result.add(attachment);
+        }
+        return result;
+    }
+
+    public void onClickAddAttachment(RecipientPresenter recipientPresenter) {
+        AttachErrorState maybeAttachErrorState = recipientPresenter.getCurrentCryptoStatus().getAttachErrorStateOrNull();
+        if (maybeAttachErrorState != null) {
+            recipientPresenter.showPgpAttachError(maybeAttachErrorState);
+            return;
+        }
+
+        attachmentMvpView.showPickAttachmentDialog(REQUEST_CODE_ATTACHMENT_URI);
+    }
+
+    private void addAttachment(Uri uri) {
+        addAttachment(uri, null);
+    }
+
+    public void addAttachment(Uri uri, String contentType) {
+        Attachment attachment = new Attachment();
+        attachment.uri = uri;
+        attachment.state = Attachment.LoadingState.URI_ONLY;
+        attachment.contentType = contentType;
+        attachment.loaderId = getNextFreeLoaderId();
+
+        if (attachments.containsKey(uri)) {
+            return;
+        }
+
+        attachments.put(uri, attachment);
+        attachmentMvpView.addAttachmentView(attachment);
+
+        initAttachmentInfoLoader(attachment);
+    }
+
+    private void initAttachmentInfoLoader(Attachment attachment) {
+        Bundle bundle = new Bundle();
+        bundle.putParcelable(LOADER_ARG_ATTACHMENT, attachment.uri);
+
+        loaderManager.initLoader(attachment.loaderId, bundle, mAttachmentInfoLoaderCallback);
+    }
+
+    private void initAttachmentContentLoader(Attachment attachment) {
+        Bundle bundle = new Bundle();
+        bundle.putParcelable(LOADER_ARG_ATTACHMENT, attachment.uri);
+
+        loaderManager.initLoader(attachment.loaderId, bundle, mAttachmentContentLoaderCallback);
+    }
+
+    private int getNextFreeLoaderId() {
+        if (nextLoaderId >= MAX_TOTAL_LOADERS) {
```

Comment:
I could not get the previous mechanism with reusable loader ids to work reliably since between all the states of the android lifecycle, there is no way to ascertain which loaders are working for which attachments at any given point. 

For that reason I simply don't reuse loader ids anymore. Having a fixed upper bound of attachments doesn't seem problematic, and it's an int namespace so the limit can be high.